tests/core/starlark/provider_tests.bzl

```diff
@@ -0,0 +1,36 @@
+load("@bazel_skylib//lib:unittest.bzl", "analysistest", "asserts")
+load("//go:def.bzl", "GoLibrary", "GoSource", "go_binary", "go_test")
+
+# The presence of these providers allows a target to be used as `deps` or `embed`.
+# go_binary and go_test targets must not be used in this way; their dependencies
+# may be built in different modes, resulting in conflicts and opaque errors.
+def _providers_test_impl(ctx):
+    env = analysistest.begin(ctx)
+    target = analysistest.target_under_test(env)
+    asserts.false(env, GoSource in target, "Should not have GoSource")
+    asserts.false(env, GoLibrary in target, "Should not have GoLibrary")
+
+    return analysistest.end(env)
+
+providers_test = analysistest.make(_providers_test_impl)
+
+def provider_test_suite():
+    go_binary(
```

Comment:
I tried to test this more directly like so:
```
def _providers_test_impl(ctx):
    env = analysistest.begin(ctx)
    asserts.expect_failure(env, "not allowed")
    return analysistest.end(env)

providers_test = analysistest.make(_providers_test_impl)

def provider_test_suite():
    go_binary(
        name = "go_binary",
        tags = ["manual"],
    )

    go_library(
        name = "binary_lib",
        deps = [":go_binary"],
        tags = ["manual"],
    )

    providers_test(
        name = "go_binary_providers_test",
        target_under_test = ":binary_lib",
    )
```

but that results in an error in analysis itself, not in test phase:
```
ERROR: /Users/dzbarsky/rules_go/tests/core/starlark/BUILD.bazel:10:20: in deps attribute of go_library rule //tests/core/starlark:binary_lib: '//tests/core/starlark:go_binary' does not have mandatory providers: 'GoLibrary'. Since this rule was created by the macro 'provider_test_suite', the error might have been caused by the macro implementation
```

any ideas how to do this better?